core/fetcher.go

```diff
@@ -160,64 +171,82 @@ func (f *BlockFetcher) ValidatorSet(ctx context.Context, height int64) (*types.V
-	f.cancel = cancel
-	f.doneCh = make(chan struct{})
-	f.isListeningForBlocks.Store(true)
+func (f *BlockFetcher) runSubscriber() (chan types.EventDataSignedBlock, error) {
+	var (
+		subscription coregrpc.BlockAPI_SubscribeNewHeightsClient
+		err          error
+	)
 
-	subscription, err := f.client.SubscribeNewHeights(ctx, &coregrpc.SubscribeNewHeightsRequest{})
-	if err != nil {
-		return nil, err
-	}
+	signedBlockCh := make(chan types.EventDataSignedBlock, 1)
 
-	signedBlockCh := make(chan types.EventDataSignedBlock)
 	go func() {
 		defer close(f.doneCh)
 		defer close(signedBlockCh)
-		defer func() { f.isListeningForBlocks.Store(false) }()
+		timeout := time.NewTimer(retrySubscriptionDelay)
+		defer timeout.Stop()
 		for {
 			select {
-			case <-ctx.Done():
+			case <-f.ctx.Done():
+				return
+			default:
+			}
+
+			subscription, err = f.client.SubscribeNewHeights(f.ctx, &coregrpc.SubscribeNewHeightsRequest{})
+			switch {
+			case err == nil:
+			case errors.Is(err, context.Canceled):
 				return
 			default:
-				resp, err := subscription.Recv()
-				if err != nil {
-					log.Errorw("fetcher: error receiving new height", "err", err.Error())
-					continue
-				}
-				withTimeout, ctxCancel := context.WithTimeout(ctx, 10*time.Second)
-				signedBlock, err := f.GetSignedBlock(withTimeout, resp.Height)
-				ctxCancel()
-				if err != nil {
-					log.Errorw("fetcher: error receiving signed block", "height", resp.Height, "err", err.Error())
-					// sleeping a bit to avoid retrying instantly and give time for the gRPC connection
-					// to recover automatically.
-					time.Sleep(time.Second)
-					continue
-				}
-				select {
-				case signedBlockCh <- types.EventDataSignedBlock{
-					Header:       *signedBlock.Header,
-					Commit:       *signedBlock.Commit,
-					ValidatorSet: *signedBlock.ValidatorSet,
-					Data:         *signedBlock.Data,
-				}:
-				case <-ctx.Done():
-					return
-				}
+				log.Errorw("fetcher: failed to subscribe to new block events", "err", err)
+				timeout.Reset(retrySubscriptionDelay)
+				<-timeout.C
+				continue
+			}
+
+			err = f.receive(f.ctx, signedBlockCh, subscription)
+			if err != nil {
+				log.Errorw("fetcher: error receiving new height", "err", err.Error())
+				continue
 			}
 		}
 	}()
-
 	return signedBlockCh, nil
 }
 
+func (f *BlockFetcher) receive(
+	ctx context.Context,
+	signedBlockCh chan types.EventDataSignedBlock,
+	subscription coregrpc.BlockAPI_SubscribeNewHeightsClient,
+) error {
+	for {
+		resp, err := subscription.Recv()
+		if err != nil {
+			return err
+		}
+
+		withTimeout, ctxCancel := context.WithTimeout(ctx, 10*time.Second)
+		signedBlock, err := f.GetSignedBlock(withTimeout, resp.Height)
+		ctxCancel()
+		if err != nil {
+			log.Errorw("fetcher: error receiving signed block", "height", resp.Height, "err", err.Error())
+			// sleeping a bit to avoid retrying instantly and give time for the gRPC connection
+			// to recover automatically.
+			time.Sleep(time.Second)
```

Comment:
Why would we need sleep, with backoff in dialer?

Reply:
did not touch this code. moved it from the previous version.